internal/coreinternal/golden/normalize_timestamps.go

```diff
@@ -0,0 +1,198 @@
+		}
+		hashCache[attrHash] = true
+	}
+}
+
+func normSummaryDataPointSlice(sdps pmetric.SummaryDataPointSlice) {
+	hashCache := make(map[[16]byte]bool)
+	for i := 0; i < sdps.Len(); i++ {
+		attrHash := pdatautil.MapHash(sdps.At(i).Attributes())
+		if hashCache[attrHash] {
+			continue
+		}
+		ts := []pcommon.Timestamp{sdps.At(i).Timestamp()}
+		for j := i + 1; j < sdps.Len(); j++ {
+			if attrHash == pdatautil.MapHash(sdps.At(j).Attributes()) {
+				ts = append(ts, sdps.At(j).Timestamp())
+			}
+		}
+
+		normalizedTs := normalizeTimestampSlice(ts)
+		for k := 0; k < sdps.Len(); k++ {
+			if pdatautil.MapHash(sdps.At(k).Attributes()) == attrHash {
+				for key, value := range normalizedTs {
+					if sdps.At(k).Timestamp() == value {
+						sdps.At(k).SetTimestamp(key)
+					}
+				}
+			}
+		}
+		hashCache[attrHash] = true
+	}
+}
+
+func normExponentialHistogramDataPointSlice(ehdps pmetric.ExponentialHistogramDataPointSlice) {
+	hashCache := make(map[[16]byte]bool)
+	for i := 0; i < ehdps.Len(); i++ {
+		attrHash := pdatautil.MapHash(ehdps.At(i).Attributes())
+		if hashCache[attrHash] {
+			continue
+		}
+		ts := []pcommon.Timestamp{ehdps.At(i).Timestamp()}
+		for j := i + 1; j < ehdps.Len(); j++ {
+			if attrHash == pdatautil.MapHash(ehdps.At(j).Attributes()) {
+				ts = append(ts, ehdps.At(j).Timestamp())
+			}
+		}
+
+		normalizedTs := normalizeTimestampSlice(ts)
+		for k := 0; k < ehdps.Len(); k++ {
+			if pdatautil.MapHash(ehdps.At(k).Attributes()) == attrHash {
+				for key, value := range normalizedTs {
+					if ehdps.At(k).Timestamp() == value {
+						ehdps.At(k).SetTimestamp(key)
+					}
+				}
+			}
+		}
+		hashCache[attrHash] = true
+	}
+}
+
+func normHistogramDataPointSlice(hdps pmetric.HistogramDataPointSlice) {
+	hashCache := make(map[[16]byte]bool)
+	for i := 0; i < hdps.Len(); i++ {
+		attrHash := pdatautil.MapHash(hdps.At(i).Attributes())
+		if hashCache[attrHash] {
+			continue
+		}
+		ts := []pcommon.Timestamp{hdps.At(i).Timestamp()}
+		for j := i + 1; j < hdps.Len(); j++ {
+			if attrHash == pdatautil.MapHash(hdps.At(j).Attributes()) {
+				ts = append(ts, hdps.At(j).Timestamp())
+			}
+		}
+
+		normalizedTs := normalizeTimestampSlice(ts)
+		for k := 0; k < hdps.Len(); k++ {
+			if pdatautil.MapHash(hdps.At(k).Attributes()) == attrHash {
+				for key, value := range normalizedTs {
+					if hdps.At(k).Timestamp() == value {
+						hdps.At(k).SetTimestamp(key)
+					}
+				}
+			}
+		}
+		hashCache[attrHash] = true
+	}
+}
+
+// sorts a slice of pcommon.Timestamp and returns a normalized slice of pcommon.Timestamp
+// normalization entails setting nonunique subsequent timestamps to the same value while incrementing unique timestamps by a set value of 1,000,000 ns
+func normalizeTimestampSlice(ts []pcommon.Timestamp) map[pcommon.Timestamp]pcommon.Timestamp {
+	normalizedTs := make(map[pcommon.Timestamp]pcommon.Timestamp)
+	tempTs := make([]pcommon.Timestamp, len(ts))
+	copy(tempTs, ts)
+	sort.Slice(tempTs, func(i, j int) bool {
```

Comment:
I don't think we need the temp slice. We created the timeseries slice and can just sort it, then set the timestamps according to the resulting index. It won't affect the order of the _data point_ slice.